src/System.Linq/src/System/Linq/Enumerable.SelectWithCountIterator.cs

```diff
@@ -0,0 +1,62 @@
+﻿
+using System.Collections.Generic;
+
+namespace System.Linq
+{
+    public static partial class Enumerable
+    {
+        internal abstract class SelectWithCountIterator<TSource> : SelectIterator<TSource>, ICollection<TSource>
+        {
+            public override abstract Iterator<TSource> Clone();
+
+            public override abstract bool MoveNext();
+
+            public override abstract IEnumerable<TSource> Where(Func<TSource, bool> predicate);
+
+            public override abstract IEnumerable<TResult> SelectImpl<TResult>(Func<TSource, TResult> selector);
+
+            #region ICollection<TSource> implementation
+
+            public abstract int Count { get; }
+
+            public bool IsReadOnly { get { return true; } }
+
+            public void Add(TSource item)
+            {
+                throw NotImplemented.ByDesign;
+            }
+
+            public void Clear()
+            {
+                throw NotImplemented.ByDesign;
+            }
+
+            public bool Contains(TSource item)
+            {
+                throw NotImplemented.ByDesign;
+            }
+
+            public void CopyTo(TSource[] array, int arrayIndex)
+            {
+                IEnumerator<TSource> enumerator = GetEnumerator();
+                int index = 0;
+				for(; index < array.Length && enumerator.MoveNext(); index++)
+                {
```

Comment:
For sources that cannot change Count while iterating (array, List), we should check that array is large enough or throw InvalidArgumentException before starting copying. 
For ICollection sources with transient Count (like ConcurrentDictionary) - I am not sure how such contract can be implemented efficiently. You can pass a scratch array to the source.CopyTo (which should be atomic WRT count changes and should validate the array size) and then, if successful, project the scratch array into given destination. That does not seem very efficient ...